### PR TITLE
Feature/condo/sberdoma 97/status model upgrade

### PR DESCRIPTION
**Pull request by Dimitreee**

### CODE REVIEW

File: apps/condo/domains/organization/gql.js [hunk 1]
```diff
@@ -1,10 +1,10 @@
 const gql = require('graphql-tag')
 const { genTestGQLUtils } = require('@core/keystone/gen.gql.utils')
 
-const ORGANIZATION_FIELDS = '{ id dv sender v country name description avatar { publicUrl } }'
-const Organization = genTestGQLUtils('Organization', ORGANIZATION_FIELDS)
+const ORGANIZATION_FIELDS = '{ id dv sender v country name description avatar { publicUrl } statusTransitions defaultEmployeeRoleStatusTransitions }'
+const Gql = genTestGQLUtils('Organization', ORGANIZATION_FIELDS)
```
Comment (pahaz) [inline apps/condo/domains/organization/gql.js:5]: ? why we need it

Comment (Dimitreee) [reply apps/condo/domains/organization/gql.js:5]: already fix it

File: apps/condo/domains/organization/schema/Organization.js [hunk 2]
```diff
@@ -9,9 +9,11 @@ const { GQLListSchema } = require('@core/keystone/schema')
 const { Json } = require('@core/keystone/fields')
 const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
 
-const { SENDER_FIELD, DV_FIELD } = require('../_common')
-const { rules } = require('../../access')
+const { SENDER_FIELD, DV_FIELD } = require('../../../schema/_common')
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/Organization.js:12]: @condo/domains/...

File: apps/condo/domains/organization/schema/Organization.js [hunk 3]
```diff
@@ -59,6 +61,38 @@ const Organization = new GQLListSchema('Organization', {
             ref: 'OrganizationEmployee.organization',
             many: true,
         },
+        statusTransitions: {
+            schemaDoc: 'Graph of possible transitions for statuses. If there is no transition in this graph, ' +
+                'it is impossible to change status if the user in the role has the right to do so.',
+            type: Json,
+            defaultValue: DEFAULT_STATUS_TRANSITIONS,
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/Organization.js:68]: isRequired: true

File: apps/condo/domains/organization/schema/Organization.js [hunk 4]
```diff
@@ -59,6 +61,38 @@ const Organization = new GQLListSchema('Organization', {
...
+                update: rules.canUpdateTicketStatusTransitions,
+                create: rules.canUpdateTicketStatusTransitions,
+                read: true,
+            },
+        },
+        defaultEmployeeRoleStatusTransitions: {
+            schemaDoc: 'Default employee role status transitions map which will be used as fallback for status transition validation' +
+                'if user dont have OrganizationEmployeeRole',
+            type: Json,
+            defaultValue: DEFAULT_STATUS_TRANSITIONS,
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/Organization.js:84]: isRequired: true

File: apps/condo/domains/ticket/constants/statusTransitions.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    'dv':1,
+    [STATUS_IDS.OPEN]: [STATUS_IDS.IN_PROGRESS, STATUS_IDS.DECLINED],
+    [STATUS_IDS.IN_PROGRESS]: [STATUS_IDS.DEFERRED, STATUS_IDS.COMPLETED, STATUS_IDS.OPEN],
+    [STATUS_IDS.COMPLETED]: [STATUS_IDS.OPEN], // TODO(Dimitreee): ADD relation to 'closed' status
+    [STATUS_IDS.DECLINED]: [],
+    [STATUS_IDS.DEFERRED]: [STATUS_IDS.OPEN],
+}
+
+module.exports = {
+    DEFAULT_STATUS_TRANSITIONS,
```
Comment (pahaz) [inline apps/condo/domains/ticket/constants/statusTransitions.js:23]: I whin it should be different for different languages.
There are Russian language statuses.

Comment (pahaz) [reply apps/condo/domains/ticket/constants/statusTransitions.js:23]: we have `organization.language`

File: apps/condo/domains/ticket/utils/status.spec.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                    expect(getPossibleStatuses(
+                        [{ id:'1' }],
+                        '1',
+                        { '2': [] }
+                    )).toStrictEqual([])
+                })
+            })
+
+
+            it('if intersections is XOR ', () => {
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/status.spec.ts:40]: It's AND. Not XOR

File: apps/condo/domains/ticket/utils/status.spec.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
+import { getPossibleStatuses } from './status'
+
+describe('status', () => {
+    describe('getPossibleStatuses', () => {
+        describe('should return correct statuses collection', () => {
+            it('if transitionsFromEmployee and transitionsFromOrganization is valid', () => {
+                expect(getPossibleStatuses(
+                    [{ id: '1' }, { id: '3' }],
+                    '1',
+                    { '1': [ '3' ] },
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/status.spec.ts:10]:
```suggestion
                    { '1': ['3'] },
```

File: apps/condo/domains/ticket/utils/status.spec.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+        describe('should return an empty statuses collection', () => {
+            describe('if transitionsFromOrganization', () => {
+                it('is not defined', () => {
+                    expect(getPossibleStatuses([], '1')).toStrictEqual([])
+                })
+
+                it('has an empty Intersection rules for status id', () => {
+                    expect(getPossibleStatuses(
+                        [{ id:'1' }],
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/status.spec.ts:24]:
```suggestion
                        [{ id: '1' }],
```

File: apps/condo/domains/ticket/utils/status.spec.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                        '1',
+                        { '1': [] }
+                    )).toStrictEqual([])
+                })
+
+                it('has no Intersection rules for status id', () => {
+                    expect(getPossibleStatuses(
+                        [{ id:'1' }],
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/status.spec.ts:32]:
```suggestion
                        [{ id: '1' }],
```

**Comment by pahaz** [comment]
What about migrations?

File: apps/condo/domains/organization/schema/Organization.js (hunk 2 above)
Comment (Dimitreee) [reply apps/condo/domains/organization/schema/Organization.js:12]: This directory is not in a domain

File: apps/condo/domains/ticket/utils/status.spec.ts (hunk 6 above)
Comment (Dimitreee) [reply apps/condo/domains/ticket/utils/status.spec.ts:40]: 'if status exists in employee perminsions or in the org permisions but not in both'

**Comment by Dimitreee** [comment]
@pahaz apps/condo/migrations/20210407112641-0009_auto_20210407_1128.js

File: apps/condo/domains/organization/schema/OrganizationEmployeeRole.js [hunk 10]
```diff
@@ -17,6 +20,23 @@ const OrganizationEmployeeRole = new GQLListSchema('OrganizationEmployeeRole', {
             isRequired: true,
         },
 
+        statusTransitions: {
+            schemaDoc: 'Employee status transitions map',
+            type: Json,
+            required: true,
+            defaultValue: DEFAULT_STATUS_TRANSITIONS,
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/OrganizationEmployeeRole.js:27]: I think the default value depends on the language. And we should create it in RegisterNewOrganization

For example here: https://github.com/open-condo-software/condo/blob/18f4689e33c79ebaa8aa0a2922d994f594b46609/apps/condo/utils/serverSchema/Organization.js#L32 we create a User Admin Role depends on the language.

I think we can put some logic here: https://github.com/open-condo-software/condo/blob/18f4689e33c79ebaa8aa0a2922d994f594b46609/apps/condo/utils/serverSchema/Organization.js#L5 and chose the right transition graph (depends on language)

**Review by pahaz — APPROVED** [review]

**Review by dkoviazin — APPROVED** [review]
Ок.

File: apps/condo/domains/organization/schema/OrganizationEmployeeRole.js (hunk 10 above)
Comment (Dimitreee) [reply apps/condo/domains/organization/schema/OrganizationEmployeeRole.js:27]: https://doma.atlassian.net/browse/SBERDOMA-118

File: apps/condo/domains/ticket/constants/statusTransitions.js (hunk 5 above)
Comment (Dimitreee) [reply apps/condo/domains/ticket/constants/statusTransitions.js:23]: https://doma.atlassian.net/browse/SBERDOMA-118

**Review by pahaz — APPROVED** [review]

**Pull request merged**